src/gennylib/include/gennylib/parallel.hpp

```diff
@@ -125,27 +149,28 @@ class AtomicContainer {
      * the iterator or other threads.
      */
     typename T::iterator begin() {
-        //std::lock_guard<std::mutex> lock(_mutex);
+        std::lock_guard<std::mutex> lock(_mutex);
         return _container.begin();
     }
+    typename T::iterator end() {
+        std::lock_guard<std::mutex> lock(_mutex);
+        return _container.end();
+    }
+
     typename T::const_iterator begin() const {
```

Comment:
Not sure if you're ready for much scrutiny here yet, but as I was scrolling by I noticed the locks here. I _think_ the lock may need to be held for the entire iteration (i.e. until `it == end()`).

I honestly don't know for sure here, but intuitively we don't want to allow new items to be added in one thread while another is iterating.

Reply:
Agreed that holding the iterator for the entire iteration would be necessary for thread safety. I'm wary about baking that into an iterator wrapper definition because the semantics might be odd. While ranged-for loops and many STL algorithms might work with an iterator that holds the lock until it's compared to `end()`, that would open the door to deadlocks if someone has multiple iterators in one thread and might be otherwise unintuitive.

Right now we're only doing iteration in single-threaded contexts, so I'd rather leave this thread-unsafe until needed. It seems intuitive that multi-operation iteration wouldn't be atomic the same way single operations are, imho? I added a comment above saying so.

That said, please push back if you disagree!

Reply:
The kicker is here:

> I'd rather leave this thread-unsafe until needed

I agree! But! I fear that if we implicitly don't solve the problem (or shrug and add a ⚠️ comment somewhere) that somebody in the future (maybe us) doesn't read or understand or _accidentally_ causes multi-threaded access during iteration. 

My suggestion:

Create an iterator wrapper not super different from what PhaseLoop already does. Have it set an "in use" flag in the collection upon `.begin()` and make the collection explicitly unset it:

```
for (const auto&& it : items) {
  do_something(it);
}
items.notify_done()
```

Then, in any member-functions that want the mutex, assert that flag isn't set and barf spectacularly if it is.

This "soft"-prevents other threads from modifying the state during iteration in a defensive-programming way, not in a compiler-ensured way.

I'll echo your sentiment to please push back. Perhaps this is being too conservative. But I personally am super conservative with doing anything "fancy" (or not strictly-necessary given the features) with concurrency in genny simply because any kind of data-race, deadlock, or generally hard-to-explain behavior will destroy a lot of trust in the utility Genny provides.

Reply:
That's a fair point. I was probably being too casual about pushing the responsibility of safety to users of the class when a defensive approach would be preferred.

I updated the class to meet the `BasicLockable` requirement and enforced that the lock is being held explicitly when iterating.